tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+							}
+						}
+
+						return nil
+					}),
+				},
+			})
+		})
+
+		t.Run("Backfilled historical events with MSC2716_HISTORICAL do not come down in an incremental sync", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			// Create the "live" event we are going to insert our backfilled events next to
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			// Create some "live" events to saturate and fill up the /sync response
+			createMessagesInRoom(t, alice, roomID, 5)
+
+			// Insert a backfilled event
+			backfillRes := backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				"",
+				1,
+				// Status
+				200,
+			)
+			_, historicalEvents := getEventsFromBatchSendResponse(t, backfillRes)
+			backfilledEvent := historicalEvents[0]
+
+			// This is just a dummy event we search for after the backfilledEvent
+			eventsAfterBackfill := createMessagesInRoom(t, alice, roomID, 1)
+			eventAfterBackfill := eventsAfterBackfill[0]
+
+			// Sync until we find the eventAfterBackfill. If we're able to see the eventAfterBackfill
+			// that occurs after the backfilledEvent without seeing eventAfterBackfill in between,
+			// we're probably safe to assume it won't sync
+			alice.SyncUntil(t, "", `{ "room": { "timeline": { "limit": 3 } } }`, "rooms.join."+client.GjsonEscape(roomID)+".timeline.events", func(r gjson.Result) bool {
+				if r.Get("event_id").Str == backfilledEvent {
+					t.Fatalf("We should not see the %s backfilled event in /sync response but it was present", backfilledEvent)
+				}
+
+				return r.Get("event_id").Str == eventAfterBackfill
+			})
+		})
+
+		t.Run("Unrecognised prev_event ID will throw an error", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+
+			backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				"$some-non-existant-event-id",
+				time.Now(),
+				"",
+				1,
+				// Status
+				// TODO: Seems like this makes more sense as a 404
+				// But the current Synapse code around unknown prev events will throw ->
+				// `403: No create event in auth events`
+				403,
+			)
+		})
+
+		t.Run("Normal users aren't allowed to backfill messages", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			backfillBatchHistoricalMessages(
+				t,
+				alice,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				"",
+				1,
```

Comment:
Given how frequently we call `backfillBatchHistoricalMessages` it's a shame that it accepts so many args that it's hard to remember what `1` and `""` are supposed to mean. This probably means this function is doing too much, but I don't know how best to break it up so it's clearer.

Reply:
Named arguments would be nice here. Code editor hinting helps for writing but not reading it. 🤷‍♀️🤔